pkg/network/node/networkpolicy.go

```diff
@@ -276,15 +276,10 @@ func (np *networkPolicyPlugin) DeleteNetNamespace(netns *osdnv1.NetNamespace) {
 	np.updateMatchCache(npns)
 }
 
-func (np *networkPolicyPlugin) SetUpPod(podIP string) error {
-	otx := np.node.oc.NewTransaction()
-	// when a pod is created isolate the pod until the current network policy rules can be
-	// evaluated
-	otx.AddFlow("table=27, priority=500, cookie=1, ip, nw_src=%s, actions=drop", podIP)
-	otx.AddFlow("table=80, priority=500, cookie=1, ip, nw_src=%s, actions=drop", podIP)
-	if err := otx.Commit(); err != nil {
-		return fmt.Errorf("Error syncing OVS flows to isolate pods %v", err)
-	}
+func (np *networkPolicyPlugin) SetUpPod(pod *corev1.Pod) error {
+	np.lock.Lock()
+	defer np.lock.Unlock()
+	np.refreshPodNetworkPolicies(pod)
```

Comment:
if you look at `refreshPodNetworkPolicies`, the only thing it uses `pod` for is to decide what namespaces need to be updated. It always gets pod IPs from the pod objects in the informer cache.

You will need to keep a separate cache of "local pod IPs that are known to openshift-sdn but not yet known to the apiserver", and then in all the places that look at `pod.Status.PodIP`, you'll have to instead use some helper that gets the IP either from the pod object (if it's set) or the cache (if it's not).

Reply:
Instead of a cache, I added some args to update network policy functions to pass podIP, trying to make less changes.
ovs flows should be updated properly now:
```
I0620 15:44:36.063846   53225 cniserver.go:297] Waiting for ADD result for pod default/hello-pod
I0620 15:44:36.063904   53225 pod.go:236] Dispatching pod network request &{ADD default hello-pod b6f490b0367c273e9c41cc90acf242acb6bf26548d0436cde2aba0f57814912d /var/run/netns/4654aa3f-1863-4bf2-b539-780ade21067a veth753f1f16  0xc0006a83c0}
I0620 15:44:36.082696   53225 ovs.go:162] Executing: ovs-vsctl --timeout=30 --may-exist add-port br0 veth753f1f16 -- set Interface veth753f1f16 external_ids=sandbox="b6f490b0367c273e9c41cc90acf242acb6bf26548d0436cde2aba0f57814912d",ip="10.128.0.65"
I0620 15:44:36.130169   53225 ovs.go:162] Executing: ovs-vsctl --timeout=30 get Interface veth753f1f16 ofport
I0620 15:44:36.154043   53225 ovs.go:160] Executing: ovs-ofctl -O OpenFlow13 bundle br0 - <<
flow add table=20, priority=100, in_port=66, arp, nw_src=10.128.0.65, arp_sha=00:00:0a:80:00:41/00:00:ff:ff:ff:ff, actions=load:0->NXM_NX_REG0[], goto_table:30
flow add table=20, priority=100, in_port=66, ip, nw_src=10.128.0.65, actions=load:0->NXM_NX_REG0[], goto_table:27
flow add table=25, priority=100, ip, nw_src=10.128.0.65, actions=load:0->NXM_NX_REG0[], goto_table:27
flow add table=40, priority=100, arp, nw_dst=10.128.0.65, actions=output:66
flow add table=70, priority=100, ip, nw_dst=10.128.0.65, actions=load:0->NXM_NX_REG1[], load:66->NXM_NX_REG2[], goto_table:80
I0620 15:44:36.211190   53225 ovs.go:162] Executing: ovs-vsctl --timeout=30 --columns=name find interface external_ids:sandbox=b6f490b0367c273e9c41cc90acf242acb6bf26548d0436cde2aba0f57814912d
I0620 15:44:36.239459   53225 ovs.go:162] Executing: ovs-vsctl --timeout=30 --if-exists clear port veth753f1f16 qos
I0620 15:44:36.264959   53225 ovs.go:162] Executing: ovs-vsctl --timeout=30 --columns=_uuid find qos external_ids:sandbox=b6f490b0367c273e9c41cc90acf242acb6bf26548d0436cde2aba0f57814912d
I0620 15:44:36.285109   53225 networkpolicy.go:886] Parsed NetworkPolicy: &node.npPolicy{policy:v1.NetworkPolicy{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"ingress-deny", GenerateName:"", Namespace:"default", SelfLink:"", UID:"423607d6-9ff5-47b3-85c2-781077dd3275", ResourceVersion:"26991", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63791336407, loc:(*time.Location)(0x31605a0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"networking.k8s.io/v1\",\"kind\":\"NetworkPolicy\",\"metadata\":{\"annotations\":{},\"name\":\"ingress-deny\",\"namespace\":\"default\"},\"spec\":{\"egress\":[{\"to\":[{\"ipBlock\":{\"cidr\":\"10.131.0.21/32\"}}]}],\"podSelector\":{\"matchLabels\":{\"name\":\"hello-pod\"}}}}\n"}, OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"kubectl-client-side-apply", Operation:"Update", APIVersion:"networking.k8s.io/v1", Time:(*v1.Time)(0xc0004da570), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0xc0004da5a0), Subresource:""}}}, Spec:v1.NetworkPolicySpec{PodSelector:v1.LabelSelector{MatchLabels:map[string]string{"name":"hello-pod"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}, Ingress:[]v1.NetworkPolicyIngressRule(nil), Egress:[]v1.NetworkPolicyEgressRule{v1.NetworkPolicyEgressRule{Ports:[]v1.NetworkPolicyPort(nil), To:[]v1.NetworkPolicyPeer{v1.NetworkPolicyPeer{PodSelector:(*v1.LabelSelector)(nil), NamespaceSelector:(*v1.LabelSelector)(nil), IPBlock:(*v1.IPBlock)(0xc000657c50)}}}}, PolicyTypes:[]v1.PolicyType{"Ingress", "Egress"}}}, watchesNamespaces:false, watchesAllPods:false, watchesOwnPods:true, selectedIPs:[]string{"10.128.0.65"}, selectsAllIPs:false, affectsIngress:true, ingressFlows:[]string(nil), affectsEgress:true, egressFlows:[]string{"ip, nw_src=10.128.0.65, ip, nw_dst=10.131.0.21/32, "}}
I0620 15:44:36.285224   53225 networkpolicy.go:374] syncNamespace 0
I0620 15:44:36.285312   53225 ovs.go:160] Executing: ovs-ofctl -O OpenFlow13 bundle br0 - <<
flow delete table=80, reg1=0
flow delete table=27, reg0=0
flow add table=80, priority=100, reg1=0, ip, nw_dst=10.128.0.65, actions=drop
flow add table=80, priority=50, reg1=0, actions=output:NXM_NX_REG2[]
flow add table=27, priority=150, reg0=0, ip, nw_src=10.128.0.65, ip, nw_dst=10.131.0.21/32,  actions=goto_table:30
flow add table=27, priority=100, reg0=0, ip, nw_src=10.128.0.65, actions=drop
flow add table=27, priority=50, reg0=0, actions=goto_table:30
I0620 15:44:36.329802   53225 pod.go:535] CNI_ADD default/hello-pod got IP 10.128.0.65, ofport 66

```

Reply:
> Instead of a cache, I added some args to update network policy functions to pass podIP, trying to make less changes.

no, you need to keep a cache, because you may get more events for that pod before its podIP gets properly set, and you don't want to be tricked into thinking the podIP got removed when you get those events

Reply:
you're right, sorry for being so stubborn, somehow cache sounded much worse than it actually is.
Hopefully this version is better